pkg/server/helpers.go

```diff
@@ -216,3 +223,129 @@ func (c *criContainerdService) getSandbox(id string) (*metadata.SandboxMetadata,
+// docker.io/library/busybox@sha256:7cc4b5aefd1d0cadf8d97d4350462ba51c694ebca145b08d7d41b41acc8db5aa.
+func normalizeImageRef(ref string) (reference.Named, error) {
+	named, err := reference.ParseNormalizedNamed(ref)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := named.(reference.NamedTagged); ok {
+		if canonical, ok := named.(reference.Canonical); ok {
+			// The reference is both tagged and digested, only
+			// return digested.
+			newNamed, err := reference.WithName(canonical.Name())
+			if err != nil {
+				return nil, err
+			}
+			newCanonical, err := reference.WithDigest(newNamed, canonical.Digest())
+			if err != nil {
+				return nil, err
+			}
+			return newCanonical, nil
+		}
+	}
+	return reference.TagNameOnly(named), nil
+}
+
+// getImageInfo returns image chainID, compressed size and oci config. Note that getImageConfig
+// assumes that the image has been pulled or it will return an error. Image reference passed-in
+// should've been normalized.
+func (c *criContainerdService) getImageInfo(ctx context.Context, ref string) (
+	imagedigest.Digest, int64, *imagespec.ImageConfig, error) {
+	normalized, err := normalizeImageRef(ref)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("failed to normalize image reference %q: %v", ref, err)
+	}
+	normalizedRef := normalized.String()
+	image, err := c.imageStoreService.Get(ctx, normalizedRef)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("failed to get image %q from containerd image store: %v",
+			normalizedRef, err)
+	}
+	// Get image config
+	desc, err := image.Config(ctx, c.contentProvider)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("failed to get image config descriptor: %v", err)
+	}
+	if desc.MediaType != imagespec.MediaTypeImageConfig && desc.MediaType != images.MediaTypeDockerSchema2Config {
+		return "", 0, nil, fmt.Errorf("unknown image config media type %q", desc.MediaType)
+	}
+	rc, err := c.contentProvider.Reader(ctx, desc.Digest)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("failed to get image config reader: %v", err)
+	}
+	defer rc.Close()
+	var imageConfig imagespec.ImageConfig
+	if err = json.NewDecoder(rc).Decode(&imageConfig); err != nil {
+		return "", 0, nil, fmt.Errorf("failed to decode image config: %v", err)
+	}
+	// Get image chainID
+	diffIDs, err := image.RootFS(ctx, c.contentProvider)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("failed to get image diff ids: %v", err)
+	}
+	chainID := identity.ChainID(diffIDs)
+	// Get image size
+	size, err := image.Size(ctx, c.contentProvider)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("failed to get image size: %v", err)
+	}
+	return chainID, size, &imageConfig, nil
+}
+
+// getRepoDigest returns image repoDigest of the named image reference.
+func getRepoDigest(namedRef reference.Named, digest imagedigest.Digest) string {
+	return namedRef.Name() + "@" + digest.String()
+}
+
+// localResolve resolves image reference to image id locally. It returns empty string
```

Comment:
s/image id/digest of the image config

Based on the implementation of this function, it can only be the digest of the image config, no? This is different from the definition of image ID at  `ImageMeta`, which is the digest of the manifest. Is this intentional?